agent/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+const (
+	metadataEnvironmentVariable = "ECS_CONTAINER_METADATA_FILE"
+	inspectContainerTimeout     = 30 * time.Second
+)
+
+type MetadataStatus string
+
+const (
+	Metadata_NOT_READY = "NOT_READY"
```

Comment:
I understand what you are saying and certainly we do not need the states stuff if we perform a single metadata file write after the container starts and we have all the data. The problem is for users to check that this file is ready in this case, they would either need a delay after container start or a file exist check after the container starts to ensure the file is valid. This puts unnecessary burden on the user to address the race condition when we can simply provide the availability of the metadata as a metadata field as well.